compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -1228,6 +1228,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// Type check a `let` statement.
     pub fn check_decl_local(&self, local: &'tcx hir::Local<'tcx>) {
         self.check_decl(local.into());
+        if let Some(blk) = local.els {
+            let previous_diverges = self.diverges.get();
+            let else_ty = self.check_block_with_expected(blk, NoExpectation);
+            let cause = self.cause(blk.span, ObligationCauseCode::LetElse);
+            if let Some(mut err) =
+                self.demand_eqtype_with_origin(&cause, self.tcx.types.never, else_ty)
+            {
+                err.emit();
+            }
+            self.diverges.set(previous_diverges);
+        }
```

Comment:
Digging into this and whether it should be part of `check_decl` I realized that `struct Let<'hir>`'s documentation still mentions `let-else`, that should be updated and its `ty` field removed.

I still think it would be better to move this into `check_decl` and extend the `Declaration` struct to also have an optional field for the else arm. 

at the same time, do make the `From<&'a hir::Local<'a>>` implementation use an exhaustive pattern instead of a `..` pattern so future changes to `Local` won't get lost